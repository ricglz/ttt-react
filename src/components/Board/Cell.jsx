import React from 'react';
import PropTypes from 'prop-types';

const currentValue = (boardGame, boardNum, cellNum) => boardGame[boardNum][cellNum];

function drawValue(boardGame, boardNum, cellNum) {
  switch (currentValue(boardGame, boardNum, cellNum)) {
    case 1:
      return 'X';
    case -1:
      return 'O';
    default:
      return '';
  }
}

const emptyFunction = () => {
  // do nothing
};

const Cell = ({
  boardGame, boardNum, cellNum, handleClick,
}) => {
  const value = drawValue(boardGame, boardNum, cellNum);
  const onClick = React.useCallback(() => {
    handleClick(boardNum, cellNum);
  }, [boardNum, cellNum, handleClick]);
  return (
    <div
      className="col-4 box"
      onClick={onClick}
      onKeyDown={emptyFunction}
      role="button"
      tabIndex="-1"
    >
      <p className={value.toLowerCase()}>{value}</p>
    </div>
  );
};

Cell.propTypes = {
  boardGame: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  boardNum: PropTypes.number.isRequired,
  cellNum: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default Cell;
