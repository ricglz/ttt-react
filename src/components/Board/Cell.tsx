import { useCallback } from "react";

import type { BigBoard, GeneralBoardIndex } from "types/general";
import { Player, Cell as CellEnum } from "types/general_enums";
import type { BoardProps } from "./Board";

type Args = {
  boardGame: BigBoard;
  boardNum: GeneralBoardIndex;
  cellNum: GeneralBoardIndex;
};

const currentValue = ({ boardGame, boardNum, cellNum }: Args) =>
  boardGame[boardNum][cellNum];

function drawValue(args: Args) {
  switch (currentValue(args)) {
    case CellEnum.X:
      return Player.PLAYER_1;
    case CellEnum.O:
      return Player.PLAYER_2;
    default:
      return "";
  }
}

const emptyFunction = () => {};

interface Props extends BoardProps {
  cellNum: GeneralBoardIndex;
}

const Cell = ({ boardGame, boardNum, cellNum, handleClick }: Props) => {
  const value = drawValue({ boardGame, boardNum, cellNum });
  const onClick = useCallback(() => {
    handleClick(boardNum, cellNum);
  }, [boardNum, cellNum, handleClick]);
  return (
    <div
      className="col-4 box"
      onClick={onClick}
      onKeyDown={emptyFunction}
      role="button"
      tabIndex={-1}
    >
      <p className={value.toLowerCase()}>{value}</p>
    </div>
  );
};

export default Cell;
