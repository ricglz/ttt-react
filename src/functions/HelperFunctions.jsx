import { NotificationManager } from 'react-notifications';

let boardCopy = null;

export function alertWinner(winner) {
  const winnerLabel = winner === 1 ? 'X' : 'O';
  NotificationManager.info(`Player ${winnerLabel} has won.`);
}

export function emptyArray() {
  return [0, 0, 0, 0, 0, 0, 0, 0, 0];
}

export function isOccupied(value) {
  return value !== 0;
}

function allThree(first, second, third) {
  const firstValue = boardCopy[first];
  const secondValue = boardCopy[second];
  const thirdValue = boardCopy[third];
  if (
    firstValue === secondValue
    && secondValue === thirdValue
    && isOccupied(firstValue)
  ) {
    return firstValue;
  }
  return null;
}

function columnWin() {
  return allThree(0, 3, 6) || allThree(1, 4, 7) || allThree(2, 5, 8);
}

function rowWin() {
  return allThree(0, 1, 2) || allThree(3, 4, 5) || allThree(6, 7, 8);
}

function diagonalWin() {
  return allThree(0, 4, 8) || allThree(2, 4, 6);
}

export function theresAWinner(board) {
  boardCopy = board;
  return columnWin() || rowWin() || diagonalWin();
}

export function newBoard() {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

export function initialState() {
  return {
    boardGame: newBoard(),
    currentPlayer: 'X',
    moveNumber: 0,
    currentBoard: -1,
  };
}

export function fbInitialState(hostUid, hostName) {
  return {
    boardGame: newBoard(),
    currentPlayer: 'X',
    moveNumber: 0,
    currentBoard: -1,
    oWins: 0,
    xWins: 0,
    hostUid,
    hostName,
    guestUid: -1,
    nextPlayerUid: hostUid,
    timestamp: Date.now(),
  };
}
