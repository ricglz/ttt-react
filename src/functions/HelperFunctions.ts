import { NotificationManager } from 'react-notifications';
import type {
  BigBoard, Board, GeneralBoardIndex, NumberBoard,
} from '../@types/general';
import { Cell, Player } from '../@types/general_enums';

export const getNextPlayer = (player: Player) => (
  player === Player.PLAYER_1 ? Player.PLAYER_2 : Player.PLAYER_1
);

export const getPlayerCellValue = (player: Player) => (
  player === Player.PLAYER_1 ? Cell.X : Cell.O
);

export function alertWinner(winner: Cell) {
  const winnerLabel = winner === Cell.X ? Player.PLAYER_1 : Player.PLAYER_2;
  NotificationManager.info(`Player ${winnerLabel} has won.`);
}

export function alertError(err: Error) {
  NotificationManager.error(err.message);
}

export function emptyArray(): NumberBoard {
  return [0, 0, 0, 0, 0, 0, 0, 0, 0];
}

export function isOccupied(value: Cell) {
  return value !== Cell.NONE;
}

function allThree(
  first: GeneralBoardIndex,
  second: GeneralBoardIndex,
  third: GeneralBoardIndex,
  board: Board,
) {
  const firstValue = board[first];
  const secondValue = board[second];
  const thirdValue = board[third];
  if (
    firstValue === secondValue
    && secondValue === thirdValue
    && isOccupied(firstValue)
  ) {
    return firstValue;
  }
  return null;
}

function columnWin(board: Board) {
  return allThree(0, 3, 6, board) || allThree(1, 4, 7, board) || allThree(2, 5, 8, board);
}

function rowWin(board: Board) {
  return allThree(0, 1, 2, board) || allThree(3, 4, 5, board) || allThree(6, 7, 8, board);
}

function diagonalWin(board: Board) {
  return allThree(0, 4, 8, board) || allThree(2, 4, 6, board);
}

export function theresAWinner(board: Board) {
  return columnWin(board) || rowWin(board) || diagonalWin(board);
}

export function newBoard(): BigBoard {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

export type CurrentBoard = GeneralBoardIndex | -1;

interface BaseGame {
  boardGame: BigBoard;
  currentPlayer: Player;
  moveNumber: number;
  currentBoard: CurrentBoard;
}

export function initialState(): BaseGame {
  return {
    boardGame: newBoard(),
    currentPlayer: Player.PLAYER_1,
    moveNumber: 0,
    currentBoard: -1,
  };
}

export interface FirebaseGame extends BaseGame {
  oWins: number;
  xWins: number;
  hostUid: string;
  hostName: string;
  guestUid: string;
  nextPlayerUid: string;
  timestamp: number;
}

export function fbInitialState(hostUid: string, hostName: string): FirebaseGame {
  return {
    ...initialState(),
    oWins: 0,
    xWins: 0,
    hostUid,
    hostName,
    guestUid: '-1',
    nextPlayerUid: hostUid,
    timestamp: Date.now(),
  };
}

export type Game = BaseGame | FirebaseGame;
