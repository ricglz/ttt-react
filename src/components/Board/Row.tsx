import React from 'react';
import { GeneralBoardIndex } from '../../functions/HelperFunctions';
import type { BoardProps } from './Board';
import Cell from './Cell';

interface Props extends BoardProps {
  rowNum: 0 | 3 | 6,
}

const Row = ({
  rowNum, boardNum, boardGame, currentBoard, handleClick,
}: Props) => (
  <div className="game row cell-row">
    <Cell
      cellNum={rowNum}
      boardNum={boardNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
    <Cell
      cellNum={rowNum + 1 as GeneralBoardIndex}
      boardNum={boardNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
    <Cell
      cellNum={rowNum + 2 as GeneralBoardIndex}
      boardNum={boardNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
  </div>
);

export default Row;
