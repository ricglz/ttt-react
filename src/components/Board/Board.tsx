import React from 'react';
import Row from './Row';
import type { CurrentBoard, GeneralBoardIndex } from '../../@types/general';
import type { BigBoardProps } from './BigBoard';

function boardClass(boardNum: GeneralBoardIndex, currentBoard: CurrentBoard) {
  let klass = 'col-4 big-box';
  if (!(boardNum === currentBoard || currentBoard === -1)) {
    klass += ' grey-bg';
  }
  return klass;
}

export interface BoardProps extends BigBoardProps {
  boardNum: GeneralBoardIndex,
}

const Board = ({
  boardNum, currentBoard, boardGame, handleClick,
}: BoardProps) => (
  <div className={boardClass(boardNum, currentBoard)}>
    <Row
      rowNum={0}
      boardNum={boardNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
    <Row
      rowNum={3}
      boardNum={boardNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
    <Row
      rowNum={6}
      boardNum={boardNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
  </div>
);

export default Board;
