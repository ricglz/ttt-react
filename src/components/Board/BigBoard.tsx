import React from 'react';
import type { BigBoard as BigBoardType, CurrentBoard } from '../../functions/HelperFunctions';
import BoardRow from './BoardRow';

type Props = {
  boardGame: BigBoardType,
  currentBoard: CurrentBoard,
  handleClick: (board: number, id: number) => void
};

const BigBoard = ({ boardGame, currentBoard, handleClick }: Props) => (
  <div className="game row justify-content-center">
    <div className="col-9 col-sm-8 col-md-7 col-lg-6 col-xl-6">
      <BoardRow
        rowNum={0}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={handleClick}
      />
      <BoardRow
        rowNum={3}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={handleClick}
      />
      <BoardRow
        rowNum={6}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={handleClick}
      />
    </div>
  </div>
);

export default BigBoard;
