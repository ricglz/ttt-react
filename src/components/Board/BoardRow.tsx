import React from 'react';
import type { GeneralBoardIndex } from '../../functions/HelperFunctions';
import type { BigBoardProps } from './BigBoard';
import Board from './Board';

interface Props extends BigBoardProps {
  rowNum: 0 | 3 | 6,
}

const BoardRow = ({
  rowNum, boardGame, currentBoard, handleClick,
}: Props) => (
  <div className="game row">
    <Board
      boardNum={rowNum}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
    <Board
      boardNum={rowNum + 1 as GeneralBoardIndex}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
    <Board
      boardNum={rowNum + 2 as GeneralBoardIndex}
      boardGame={boardGame}
      currentBoard={currentBoard}
      handleClick={handleClick}
    />
  </div>
);

export default BoardRow;
