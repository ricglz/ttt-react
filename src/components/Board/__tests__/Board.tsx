import React from 'react';
import renderer from 'react-test-renderer';
import type { CurrentBoard } from '../../../@types/general';
import { newBoard } from '../../../functions/HelperFunctions';
import Board from '../Board';

describe('Board', () => {
  const boardNum = 0;
  let currentBoard: CurrentBoard;

  afterEach(() => {
    const boardGame = newBoard();
    const component = (
      <Board
        boardNum={boardNum}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={() => {}}
      />
    );
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  })

  describe('currentBoard value', () => {
    test('currentBoard === -1', () => {
      currentBoard = -1;
    });

    test('currentBoard === boardNum', () => {
      currentBoard = boardNum;
    });

    test('currentBoard !== boardNum', () => {
      currentBoard = boardNum + 1 as CurrentBoard;
    });
  });
})
