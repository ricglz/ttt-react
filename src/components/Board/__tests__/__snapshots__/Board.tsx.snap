// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board currentBoard value currentBoard !== boardNum 1`] = `
<div
  className="col-4 big-box grey-bg"
>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
</div>
`;

exports[`Board currentBoard value currentBoard === -1 1`] = `
<div
  className="col-4 big-box"
>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
</div>
`;

exports[`Board currentBoard value currentBoard === boardNum 1`] = `
<div
  className="col-4 big-box"
>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
  <div
    className="game row cell-row"
  >
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
    <div
      className="col-4 box"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <p
        className=""
      >
        
      </p>
    </div>
  </div>
</div>
`;
